libs/shared/src/main/java/io/crate/common/collections/Maps.java

```diff
@@ -155,4 +158,37 @@ public static void mergeInto(Map<String, Object> source,
         }
         return root;
     }
+
+
+    /**
+     * Add entries in `additions` to `map`.
+     * If `map` already contains an entry that is also present in `addition`:
+     *  - It will recurse into it if it is a map
+     *  - Skip the entry otherwise
+     **/
+    public static void extendRecursive(Map<String, Object> map, Map<String, Object> additions) {
+        extendRecursive(map, additions, (oldList, newList) -> Lists2.concatUnique(oldList, newList));
+    }
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    public static void extendRecursive(Map<String, Object> map,
```

Comment:
could we use this also as a replacement for `TransportSchemaUpdateAction.mergeIntoSource(...)`?

Reply:
I thought so at first, but `TransportSchemaUpdateAction` has some validation integrated:

```
                    if (updateAllowed(key, sourceValue, updateValue)) {
                        source.put(key, updateValue);
                    } else if (!isUpdateIgnored(path) && !sourceValue.equals(updateValue)) {
                        String fqKey = String.join(".", path) + '.' + key;
                        throw new IllegalArgumentException(
                            "Can't overwrite " + fqKey + "=" + sourceValue + " with " + updateValue);
                    }
```

This is also the reason it additionally keeps track of the `path`. Didn't feel like it is worth to generalize that requirement.

Reply:
ah right 👍